bindings/matrix-sdk-ffi/src/notification_settings.rs

```diff
@@ -0,0 +1,213 @@
+    }
+}
+
+#[derive(Clone, uniffi::Object)]
+pub struct NotificationSettings {
+    pub(crate) sdk_client: MatrixClient,
+    push_rules: Arc<RwLock<Ruleset>>,
+    delegate: Arc<RwLock<Option<Box<dyn NotificationSettingsDelegate>>>>,
+}
+
+impl NotificationSettings {
+    pub(crate) fn new(sdk_client: MatrixClient) -> Self {
+        let push_rules = Arc::new(tokio::sync::RwLock::new(Ruleset::new()));
+        let delegate: Arc<RwLock<Option<Box<dyn NotificationSettingsDelegate>>>> =
+            Arc::new(RwLock::new(None));
+
+        // Listen for PushRulesEvent
+        let push_rules_clone = push_rules.to_owned();
+        let delegate_clone = delegate.to_owned();
+        sdk_client.add_event_handler(move |ev: PushRulesEvent| {
+            let push_rules = push_rules_clone.clone();
+            let delegate = delegate_clone.clone();
+            async move {
+                *push_rules.write().await = ev.content.global;
+                if let Some(delegate) = delegate.read().await.as_ref() {
+                    delegate.notification_settings_did_change();
+                }
+            }
+        });
+
+        Self { sdk_client, push_rules, delegate }
+    }
+}
+
+#[uniffi::export(async_runtime = "tokio")]
+impl NotificationSettings {
+    /// Sets a delegate.
+    pub async fn set_delegate(&self, delegate: Option<Box<dyn NotificationSettingsDelegate>>) {
+        *self.delegate.write().await = delegate;
+    }
+
+    /// Gets the notification mode for a room.
+    ///
+    /// # Arguments
+    ///
+    /// * `room_id` - A room ID
+    pub async fn get_room_notification_mode(
+        &self,
+        room_id: String,
+    ) -> Result<RoomNotificationSettings, NotificationSettingsError> {
+        let ruleset = &*self.push_rules.read().await;
+        let notification_settings = self.sdk_client.notification_settings();
+        let parsed_room_id = RoomId::parse(&room_id)
+            .map_err(|_e| NotificationSettingsError::InvalidRoomId(room_id))?;
+        // Get the current user defined mode for this room
+        if let Some(mode) =
+            notification_settings.get_user_defined_room_notification_mode(&parsed_room_id, ruleset)
+        {
+            return Ok(RoomNotificationSettings::new(mode.into(), false));
+        }
+
+        // If the user didn't defined a notification mode, return the default one for
+        // this room
+        let room = self
+            .sdk_client
+            .get_room(&parsed_room_id)
+            .context("Room not found")
+            .map_err(|_| NotificationSettingsError::RoomNotFound)?;
+
+        let is_encrypted = room.is_encrypted().await.unwrap_or(false);
+        let members_count = room.joined_members_count();
+
+        let mode = notification_settings.get_default_room_notification_mode(
+            is_encrypted,
+            members_count,
+            ruleset,
+        );
+        Ok(RoomNotificationSettings::new(mode.into(), true))
```

Comment:
IMO this should be a method in the `NotificationSettings` itself, there's still a bit too much logic for the FFI layer here.